api/src/main/java/org/pmiops/workbench/notebooks/LeonardoNotebooksClientImpl.java

```diff
@@ -97,39 +112,72 @@ private LeonardoCreateRuntimeRequest buildCreateRuntimeRequest(
-        .addScopesItem("https://www.googleapis.com/auth/cloud-platform")
-        .addScopesItem("https://www.googleapis.com/auth/userinfo.email")
-        .addScopesItem("https://www.googleapis.com/auth/userinfo.profile")
-        .runtimeConfig(
-            new LeonardoMachineConfig()
-                .cloudService(CloudServiceEnum.DATAPROC)
-                .masterDiskSize(
-                    Optional.ofNullable(clusterOverride.masterDiskSize)
-                        .orElse(config.firecloud.notebookRuntimeDefaultDiskSizeGb))
-                .masterMachineType(
-                    Optional.ofNullable(clusterOverride.machineType)
-                        .orElse(config.firecloud.notebookRuntimeDefaultMachineType)))
-        .toolDockerImage(workbenchConfigProvider.get().firecloud.jupyterDockerImage)
-        .welderDockerImage(workbenchConfigProvider.get().firecloud.welderDockerImage)
-        .customEnvironmentVariables(customEnvironmentVariables);
+    Map<String, String> runtimeLabels = new HashMap<>();
+    runtimeLabels.put(RUNTIME_LABEL_AOU, "true");
+    runtimeLabels.put(RUNTIME_LABEL_CREATED_BY, userEmail);
+
+    if (runtime.getConfigurationType() != null) {
+      runtimeLabels.put(
+          RUNTIME_LABEL_AOU_CONFIG,
+          RUNTIME_CONFIGURATION_TYPE_ENUM_TO_STORAGE_MAP.get(runtime.getConfigurationType()));
+    }
+
+    LeonardoCreateRuntimeRequest request =
+        new LeonardoCreateRuntimeRequest()
+            .labels(runtimeLabels)
+            .defaultClientId(config.server.oauthClientId)
+            // Note: Filenames must be kept in sync with files in api/src/main/webapp/static.
+            .jupyterUserScriptUri(assetsBaseUrl + "/initialize_notebook_runtime.sh")
+            .jupyterStartUserScriptUri(assetsBaseUrl + "/start_notebook_runtime.sh")
+            .userJupyterExtensionConfig(
+                new LeonardoUserJupyterExtensionConfig().nbExtensions(nbExtensions))
+            // Matches Terra UI's scopes, see RW-3531 for rationale.
+            .addScopesItem("https://www.googleapis.com/auth/cloud-platform")
+            .addScopesItem("https://www.googleapis.com/auth/userinfo.email")
+            .addScopesItem("https://www.googleapis.com/auth/userinfo.profile")
+            .toolDockerImage(workbenchConfigProvider.get().firecloud.jupyterDockerImage)
+            .welderDockerImage(workbenchConfigProvider.get().firecloud.welderDockerImage)
+            .customEnvironmentVariables(customEnvironmentVariables);
+
+    if (workbenchConfigProvider.get().featureFlags.enableCustomRuntimes) {
+      if (runtime.getGceConfig() == null && runtime.getDataprocConfig() == null) {
+        // There's probably a code smell here where I'm throwing an exception that we're only
+        // expecting to handle as a response from Leo but this was the quickest path forward for now.
+        throw new ApiException(400, "Either a GceConfig or DataprocConfig must be provided");
```

Comment:
nit: BadRequestException

Reply:
Tried that first but it doesn't work because the LeoRetryHandler tries to cast it into an `ApiException`. I think the problem here is that LeoRetryHandler has the assumption that its basically just a shim for calls to Leo APIs but we've added enough logic here that we can throw our own exceptions.